test/blackbox-tests/test-cases/tests-locks.t/run.t

```diff
@@ -0,0 +1,13 @@
+We want to test properties about tests. We need the INSIDE_DUNE to be off.
+However we always provide --root so these tests should not escape their scope.
+  $ unset INSIDE_DUNE
```

Comment:
Could we pass `-j 2` explicitly to Dune rather than clearing `INSIDE_DUNE`? Without it, there are many things that can influence the behaviour of Dune, such as the user configuration file.

Reply:
I guess this could work but there is an other issue: the diff of the inner test is not showing when `INSIDE_DUNE` is on:

The module `Print_diff` contains the following:
https://github.com/ocaml/dune/blob/d6e14f9d3b6126f5f6de44d582ac6bdd6e4f2901/src/dune_engine/print_diff.ml#L85-L86

Do you know what this restriction is for and how I could bypass it ?

Reply:
(I will also try an other strategy for testing)

Reply:
Ok I have a new design for the test which works with `INSIDE_DUNE`: I write the test files at test time predicting the exact time they will finish if they are locks or not. This assumes Dune will always runs sequential tests by alphabetical order:

```
These tests are run with locks. They should end at 1 second apart

  $ mkdir tests/subdir

  $ SEC=$(date +%s)
  $ cat >tests/a.t <<EOF
  > $ sleep 1 && date +%s
  > $(($SEC + 1))
  > EOF

  $ cat >tests/b.t <<EOF
  > $ sleep 1 && date +%s
  > $(($SEC + 2))
  > EOF

  $ dune build --root=. -j 2 @all-tests 2>&1

These tests are run without locks. They should end together
at the exact same second

  $ mkdir tests-no-locks/subdir

  $ SEC=$(date +%s)
  $ cat >tests-no-locks/c.t <<EOF
  > $ sleep 1 && date +%s
  > $(($SEC + 1))
  > EOF

  $ cat >tests-no-locks/d.t <<EOF
  > $ sleep 1 && date +%s
  > $(($SEC + 1))
  > EOF

  $ dune build --root=. -j 2 @all-tests-nolocks
  ```

It does look even more fragile than before however...

Reply:
For the diff, you can use `--diff-command`. Which is better for reproducibility, as otherwise Dune will prefer `patdiff` if it is installed.

Reply:
Makes sense, I will use that. Hopefully `diff` behaves the same on Linux and Macos.

Reply:
If `diff` causes us trouble, we could add a `diff` command to `dune_cmd` (in test/blackbox-tests/utils).

Reply:
I just wrote a test that relies on the execution order as well in #4435. Could you have a look and see if the same method I used could apply here?

Reply:
in particular, it should be reliable and doesn't rely on waiting.

Reply:
I settled with test not relying on execution order. I simply compare their ending time: tests started simultaneously should also end simultaneously while tests using locks should not.